packages/webchat/src/core/socket.tsx

```diff
@@ -1,73 +1,95 @@
-    this.events = bp?.events
-    this.userIdScope = config.userIdScope
+  constructor(config: Config) {
     this.chatId = config.chatId
+    this.socket = new MessagingSocket({ url: config.url, clientId: config.clientId })
   }
 
   public setup() {
-    if (!this.events) {
-      return
-    }
-
-    // Connect the Botpress Web Socket to the server
-    this.events.setup(this.userIdScope)
+    this.socket.on('message', this.onMessage)
 
+    /*
     this.events.on('guest.webchat.clear', this.onClear)
     this.events.on('guest.webchat.message', this.onMessage)
     this.events.on('guest.webchat.typing', this.onTyping)
     this.events.on('guest.webchat.data', this.onData)
+    */
 
     // firehose events to parent page
-    this.events.onAny(this.postToParent)
+    // TODO: why do we need this
+    // this.events.onAny(this.postToParent)
+  }
+
+  public async sendPayload(payload: any): Promise<Message> {
+    const message = await this.socket.sendPayload(payload)
+    this.onMessage(message)
+    return message
   }
 
   public postToParent = (_type: string, payload: any) => {
     // we could filter on event type if necessary
     window.parent?.postMessage({ ...payload, chatId: this.chatId }, '*')
   }
 
-  public changeUserId(newId: string) {
-    if (typeof newId === 'string' && newId !== 'undefined') {
-      this.events.updateVisitorId(newId, this.userIdScope)
-    }
-  }
-
   /** Waits until the VISITOR ID and VISITOR SOCKET ID is set  */
   public async waitForUserId(): Promise<void> {
-    return new Promise((resolve, reject) => {
-      const interval = setInterval(() => {
-        if (isString(window.__BP_VISITOR_ID) && isString(window.__BP_VISITOR_SOCKET_ID)) {
-          clearInterval(interval)
-
-          const userId = window.__BP_VISITOR_ID
-          this.onUserIdChanged(userId)
-          this.postToParent('', { userId })
-
-          resolve()
-        }
-      }, 250)
-
-      setTimeout(() => {
-        clearInterval(interval)
-        reject('Timeout to acquire VISITOR ID and VISITOR SOCKET ID exceeded.')
-      }, 30000)
+    if (this.waitingForUser) {
+      return this.waitingForUser
+    }
+
+    this.waitingForUser = new Promise<void>(async (resolve, reject) => {
+      const creds = this.getStorage<UserCredentials>('creds')
+      await this.socket.connect({ autoLogin: false })
+      await this.socket.login(creds)
+
+      if (this.socket.userId) {
+        const userId = this.socket.userId!
+        this.setStorage('creds', this.socket.creds)
+
+        this.onUserIdChanged(userId)
+        this.postToParent('', { userId })
+
+        resolve()
+      } else {
+        this.waitingForUser = undefined
+        reject()
+      }
     })
+
+    return this.waitingForUser
   }
-}
 
-const isString = (str: string | any): str is string => {
-  return typeof str === 'string' && str !== 'undefined'
+  public getStorage<T>(key: string): T | undefined {
+    const stored = localStorage.getItem(this.getStorageKey(key))
```

Comment:
Please copy the `storage.ts` file (from ui-shared-lite) instead of implementing a custom version. It supports localStorage, cookies and sessionStorage, and automatically serialize/deserialize infromations.

In some cases the local storage is disabled as a security measure.

There is also a configuration option to use sessionStorage instead of local